### PR TITLE
[nan-306] log update

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
Update nango.log to accept a recursive object

## Issue ticket number and link
NAN-306

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/shared/lib/sdk/sync.ts [hunk 1]
```diff
@@ -417,9 +412,24 @@ export class NangoAction {
...
 
-    public async log(content: string, userDefinedLevel?: UserLogParameters): Promise<void> {
+    /**
+     * Log
+     * @desc Log a message to the activity log which shows up in the Nango Dashboard
+     * note that the last argument can be an object with a level property to specify the log level
+     * example: await nango.log('This is a log message', { level: 'error' })
+     */
+    public async log(...args: any[]): Promise<void> {
+        const lastArg = args[args.length - 1];
```
Comment (TBonnin) [inline packages/shared/lib/sdk/sync.ts:437]: what happens if log is called without arguments?

File: packages/shared/lib/sdk/sync.ts [hunk 2]
```diff
@@ -417,9 +412,24 @@ export class NangoAction {
...
+        const userDefinedLevel = typeof lastArg === 'object' && lastArg.level ? lastArg : undefined;
```
Comment (TBonnin) [inline packages/shared/lib/sdk/sync.ts]: you can write a function with a type guard. something like this:
```
const isUserLogParameters = (object: any): object is UserLogParameters => {
    return 'level' in object;
}
 const logParameters = isUserLogParameters(lastArg)? lastArg : undefined;
```

File: packages/shared/lib/sdk/sync.ts [hunk 3]
```diff
@@ -417,9 +412,24 @@ export class NangoAction {
...
+
+        if (userDefinedLevel) {
+            args.pop();
+        }
+
+        const content = args.map((arg) => (typeof arg === 'object' ? safeStringify(arg) : String(arg))).join(' ');
```
Comment (TBonnin) [inline packages/shared/lib/sdk/sync.ts]: I would put the type check within safeStringify and handle all types, not just object but string, number, etc... in safeStringify so one can use it without thinking too much

File: packages/shared/lib/sdk/sync.ts [hunk 4]
```diff
@@ -437,8 +447,9 @@ export class NangoAction {
                     msg: content
                 }
             });
+
             if (response.status > 299) {
-                console.log(
+                console.error(
```
Comment (TBonnin) [inline packages/shared/lib/sdk/sync.ts:472]: I think I have copy-pasted this everywhere it is using the persistAPI so you might want to fix it also there

File: packages/shared/lib/sdk/sync.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/shared/lib/sdk/sync.ts:437]: Error throw with this update: https://github.com/NangoHQ/nango/pull/1601/commits/ceb892eaaedb1c680a2f2332cd9325fa1999b60d#diff-f52f0d61cd6802499f5a31aca89d3aec3df278a751065390c02b6bbdb24c0840R427

File: packages/shared/lib/sdk/sync.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/shared/lib/sdk/sync.ts]: Thanks, added with https://github.com/NangoHQ/nango/pull/1601/commits/ceb892eaaedb1c680a2f2332cd9325fa1999b60d#diff-f52f0d61cd6802499f5a31aca89d3aec3df278a751065390c02b6bbdb24c0840R433

File: packages/shared/lib/sdk/sync.ts (hunk 3 above)
Comment (khaliqgant) [reply packages/shared/lib/sdk/sync.ts]: Unsure what you mean by adding all types but made an update here: https://github.com/NangoHQ/nango/pull/1601/commits/ceb892eaaedb1c680a2f2332cd9325fa1999b60d#diff-f8e7cae5cc6251960bb9e297501fa4e040c7266b762c60c906980b07293abe70R393

File: packages/shared/lib/sdk/sync.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/shared/lib/sdk/sync.ts:472]: Updated with https://github.com/NangoHQ/nango/pull/1601/commits/da1f2d1b684fcf484a4640d7cd2f27cc7cdefbb7

File: packages/shared/lib/sdk/sync.ts (hunk 1 above)
Comment (TBonnin) [reply packages/shared/lib/sdk/sync.ts:437]: `console.log` prints an empty string when passing no argument. Should we do the same and have log to be a no op when no arg is passed

**Review by TBonnin — APPROVED** [review]

File: packages/shared/lib/sdk/sync.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/shared/lib/sdk/sync.ts:437]: Sure, updated with https://github.com/NangoHQ/nango/pull/1601/commits/b0734a23f5571cfccebbf9bc225bdb2adf487c86

**Pull request merged**